lib/include/currentSourceModels.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+// Standard includes
+#include <array>
+#include <functional>
+#include <string>
+#include <tuple>
+#include <vector>
+#include <cmath>
+
+// GeNN includes
+#include "codeGenUtils.h"
+// #include "neuronModels.h"
+#include "newModels.h"
+
+//----------------------------------------------------------------------------
+// Macros
+//----------------------------------------------------------------------------
+#define SET_INJECTION_CODE(INJECTION_CODE) virtual std::string getInjectionCode() const{ return INJECTION_CODE; }
+#define SET_TIME_CONDITION_CODE(TIME_CONDITION_CODE) virtual std::string getTimeConditionCode() const{ return TIME_CONDITION_CODE; }
+#define SET_EXTRA_GLOBAL_PARAMS(...) virtual StringPairVec getExtraGlobalParams() const{ return __VA_ARGS__; }
+
+//----------------------------------------------------------------------------
+// CurrentSourceModels::Base
+//----------------------------------------------------------------------------
+namespace CurrentSourceModels
+{
+//! Base class for all current source models
+class Base : public NewModels::Base
+{
+public:
+    //----------------------------------------------------------------------------
+    // Declared virtuals
+    //----------------------------------------------------------------------------
+    //! Gets the code that defines current injected each timestep 
+    virtual std::string getInjectionCode() const{ return ""; }
+
+    //! Gets code which defines the condition when the should be applied described current source model.
+    /*! This evaluates to a bool (e.g. "t > 20 && t < 40"). */
+    virtual std::string getTimeConditionCode() const{ return ""; }
```

Comment:
I don't think the time condition needs to be separate - I don't think it gains anything over including ``if($(t) > $(startTime) && $(t) <= $(endTime))`` in the injection code.

Reply:
This would require that the injection code wasn't a single statement, but I think that's nicer anyway i.e.:
```c++
if($(t) > $(tStart) && t < $(tStop)) {
    $(injectCurrent, $(amp) * sin($(omega)*$(t) + $(phase)) + $(offset));
}
```
You can then substitute the newer 'function' syntax with (you can also directly inject into ``Isyn``):
```c++
functionSubstitute(code, "injectCurrent", 1, "Isyn += $(0)");
```